### PR TITLE
Custom flat model activation

### PR DESCRIPTION
**Pull request by BadLuckBurt**
Pull request as requested (all puns intended :P)

### CODE REVIEW

File: Assets/Scripts/Utility/RDBLayout.cs [hunk 1]
```diff
@@ -626,7 +626,7 @@ public static void AddLights(GameObject go, ref DFBlock blockData)
                         }
 
                         // Check if model has an action record
-                        bool hasAction = HasAction(obj);
+                        bool hasAction = HasAction(obj, ref blockData);
```
Comment (ajrb) [inline Assets/Scripts/Utility/RDBLayout.cs]: I would prefer a separate check done where it decides to combine model or not. Also custom activations should not change the parent used should they? That's what this is doing from what I can see. Either way I don't like this change.

I would like it be simplified like this please:
`
if (combiner == null || hasAction || PlayerActivate.HasCustomActivation(modelId))
`

File: Assets/Scripts/Utility/RDBLayout.cs [hunk 2]
```diff
@@ -743,11 +743,16 @@ private static bool IsActionDoor(ref DFBlock blockData, DFBlock.RdbObject obj, i
         /// <summary>
         /// Check is model has action record.
         /// </summary>
-        private static bool HasAction(DFBlock.RdbObject obj)
+        private static bool HasAction(DFBlock.RdbObject obj, ref DFBlock blockData)
```
Comment (ajrb) [inline Assets/Scripts/Utility/RDBLayout.cs]: revert these changes, see previous comment.

File: Assets/Scripts/Game/PlayerActivate.cs [hunk 3]
```diff
@@ -281,10 +310,11 @@ void Update()
                     int pos = modelName.IndexOf(']');
                     if (pos > 0 && pos < modelName.Length - 1)
                         modelName = modelName.Remove(pos + 1);
-                    ModelActivation activation;
-                    if (customModelActivations.TryGetValue(modelName, out activation))
+
```
[the remark this answers is not in the document]
Comment (BadLuckBurt) [reply Assets/Scripts/Game/PlayerActivate.cs]: Yes, it's potentially a flat too. I propose we stick to flatModelName then as go (GameObject) may be too generic and I've used flatModel in the other changed parts already.

File: Assets/Scripts/Game/PlayerActivate.cs [hunk 4]
```diff
@@ -91,22 +91,51 @@ public static bool IsBuildingOpen(DFLocation.BuildingTypes buildingType)
...
+        public static bool RegisterCustomActivation(int textureArchive, int textureRecord, CustomActivation customActivation)
+        {
+            string goFlatName = GameObjectHelper.GetGoFlatName(textureArchive, textureRecord);
+            return HandleRegisterCustomActivation(goFlatName, customActivation);
+        }
+
+        private static bool HandleRegisterCustomActivation(string goFlatModelName, CustomActivation customActivation)
+        {
+            DaggerfallUnity.LogMessage("HandleRegisterCustomActivation: " + goFlatModelName, true);
+            if (!CheckCustomActivation(goFlatModelName))
```
[the remark this answers is not in the document]
Comment (BadLuckBurt) [reply Assets/Scripts/Game/PlayerActivate.cs]: I'll make the change to have it overwrite any existing custom activations for the goFlatModelName key but I don't see any harm in keeping this method around, otherwise both Register methods end up doing the same thing and this way, if any changes are needed or logic needs to be brought back in, it only needs to be done once.

File: Assets/Scripts/Utility/GameObjectHelper.cs [hunk 5]
```diff
@@ -268,12 +273,23 @@ public static void ChangeDaggerfallMeshGameObject(DaggerfallMesh dfMesh, uint ne
...
         {
-            GameObject go = new GameObject(string.Format("DaggerfallBillboard [TEXTURE.{0:000}, Index={1}]", archive, record));
+            string flatName = GetGoFlatName(archive, record);
+            GameObject go = new GameObject(flatName);
             if (parent) go.transform.parent = parent;
 
             DaggerfallBillboard dfBillboard = go.AddComponent<DaggerfallBillboard>();
             dfBillboard.SetMaterial(archive, record);
 
+            if(PlayerActivate.CheckCustomActivation(flatName)) {
```
[the remark this answers is not in the document]
Comment (BadLuckBurt) [reply Assets/Scripts/Utility/GameObjectHelper.cs]: Slipped up there, did it pretty consistently for the other parts where I changed stuff. Will take care of it

File: Assets/Scripts/Utility/AssetInjection/MeshReplacement.cs [hunk 6]
```diff
@@ -170,13 +175,13 @@ public static GameObject SwapCustomFlatGameobject(int archive, int record, Trans
...
 
-            string name = string.Format("DaggerfallBillboard [Replacement] [TEXTURE.{0:000}, Index={1}]", archive, record);
+            string name = GetFlatReplacementName(archive, record);
             for (int i = 0; i < parent.childCount; i++)
             {
                 Transform transform = parent.GetChild(i);
                 if (transform.name == name)
                     AlignToBase((go = transform.gameObject).transform, position, archive, record, inDungeon);
-                else if (transform.name.StartsWith("DaggerfallBillboard [Replacement]"))
+                else if (transform.name.EndsWith(" [Replacement]"))
```
[the remark this answers is not in the document]
Comment (BadLuckBurt) [reply Assets/Scripts/Utility/AssetInjection/MeshReplacement.cs]: That method (SwapCustomFlatGameobject) only deals with flats so it should be fine. I'll test it before putting in a final pull request with Interkarma.

File: Assets/Scripts/Utility/RDBLayout.cs (hunk 1 above)
Comment (BadLuckBurt) [reply Assets/Scripts/Utility/RDBLayout.cs]: Discussed on Discord and I modified it as requested. Will test again before final pull request to IK.

File: Assets/Scripts/Utility/RDBLayout.cs (hunk 2 above)
Comment (BadLuckBurt) [reply Assets/Scripts/Utility/RDBLayout.cs]: As with previous comment, discussed on Discord and I modified it as requested. Will test again before final pull request to IK.

File: Assets/Scripts/Utility/RDBLayout.cs [hunk 7]
```diff
@@ -250,7 +250,7 @@ public static void AddActionDoors(GameObject go, Dictionary<int, ActionLink> act
                                 dfMesh.SetDungeonTextures(textureTable);
 
                             // Add action component to door if it also has an action
-                            if (HasAction(obj))
+                            if (HasAction(obj, ref blockData))
```
[the remark this answers is not in the document]
Comment (BadLuckBurt) [reply Assets/Scripts/Utility/RDBLayout.cs]: Change has been reverted. Will test before final PR

File: Assets/Scripts/Game/PlayerActivate.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ajrb) [reply Assets/Scripts/Game/PlayerActivate.cs]: This is still wrong, as Add() will throw an ArgumentException if entry is already in there

File: Assets/Scripts/Game/PlayerActivate.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ajrb) [reply Assets/Scripts/Game/PlayerActivate.cs]: Okay please make the change then.

File: Assets/Scripts/Game/PlayerActivate.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (BadLuckBurt) [reply Assets/Scripts/Game/PlayerActivate.cs]: Changed PlayerActivate.HandleRegisterCustomActivation from bool to void, it now overwrites any previously registered custom actions for a given goFlatModelName.

File: Assets/Scripts/Game/PlayerActivate.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (BadLuckBurt) [reply Assets/Scripts/Game/PlayerActivate.cs]: Changed variable modelName to flatModelName in PlayerActivate.Update to better reflect what it will match from now on.

File: Assets/Scripts/Utility/GameObjectHelper.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (BadLuckBurt) [reply Assets/Scripts/Utility/GameObjectHelper.cs]: Corrected syntax (space and curly brace position) in GameObjectHelper.cs

File: Assets/Scripts/Utility/RDBLayout.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (BadLuckBurt) [reply Assets/Scripts/Utility/RDBLayout.cs]: Reverted RDBLayout.HasAction changes back to original form and added separate check to prevent models with custom actions from being combined.

File: Assets/Scripts/Utility/RDBLayout.cs (hunk 1 above)
Comment (BadLuckBurt) [reply Assets/Scripts/Utility/RDBLayout.cs]: Reverted RDBLayout.HasAction changes back to original form and added separate check to prevent models with custom actions from being combined.

File: Assets/Scripts/Utility/RDBLayout.cs (hunk 2 above)
Comment (BadLuckBurt) [reply Assets/Scripts/Utility/RDBLayout.cs]: Reverted RDBLayout.HasAction changes back to original form and added separate check to prevent models with custom actions from being combined.

File: Assets/Scripts/Game/PlayerActivate.cs [hunk 8]
```diff
@@ -91,20 +91,44 @@ public static bool IsBuildingOpen(DFLocation.BuildingTypes buildingType)
...
             string goModelName = GameObjectHelper.GetGoModelName(modelID);
-            DaggerfallUnity.LogMessage("RegisterModelActivation: " + goModelName, true);
-            if (!customModelActivations.ContainsKey(goModelName))
-            {
-                customModelActivations.Add(goModelName, modelActivation);
-                return true;
-            }
-            return false;
+            HandleRegisterCustomActivation(goModelName, customActivation);
+            return true;
```
Comment (ajrb) [inline Assets/Scripts/Game/PlayerActivate.cs]: returning true is now pointless

File: Assets/Scripts/Game/PlayerActivate.cs [hunk 9]
```diff
@@ -91,20 +91,44 @@ public static bool IsBuildingOpen(DFLocation.BuildingTypes buildingType)
...
+        }
+
+        public static bool RegisterCustomActivation(int textureArchive, int textureRecord, CustomActivation customActivation)
+        {
+            string goFlatName = GameObjectHelper.GetGoFlatName(textureArchive, textureRecord);
+            HandleRegisterCustomActivation(goFlatName, customActivation);
+            return true;
```
Comment (ajrb) [inline Assets/Scripts/Game/PlayerActivate.cs]: pointless return

File: Assets/Scripts/Game/PlayerActivate.cs [hunk 10]
```diff
@@ -91,20 +91,44 @@ public static bool IsBuildingOpen(DFLocation.BuildingTypes buildingType)
...
+            return CheckCustomActivation(goModelName);
+        }
+
+        public static bool HasCustomActivation(int textureArchive, int textureRecord)
+        {
+            string goFlatName = GameObjectHelper.GetGoFlatName(textureArchive, textureRecord);
+            return CheckCustomActivation(goFlatName);
+        }
+
+        public static bool CheckCustomActivation(string goFlatModelName) {
```
Comment (ajrb) [inline Assets/Scripts/Game/PlayerActivate.cs]: Is this method really still worth it? I think I'd just copy the containskey into the previous 2 Has.. methods.

File: Assets/Scripts/Utility/RDBLayout.cs [hunk 11]
```diff
@@ -746,8 +746,9 @@ private static bool IsActionDoor(ref DFBlock blockData, DFBlock.RdbObject obj, i
         private static bool HasAction(DFBlock.RdbObject obj)
         {
             DFBlock.RdbActionResource action = obj.Resources.ModelResource.ActionResource;
-            if (action.Flags != 0)
+            if (action.Flags != 0) {
```
Comment (ajrb) [inline Assets/Scripts/Utility/RDBLayout.cs]: If you're gonna change this, then lets make it 

`return actions.Flags != 0;

File: Assets/Scripts/Game/PlayerActivate.cs (hunk 8 above)
Comment (BadLuckBurt) [reply Assets/Scripts/Game/PlayerActivate.cs]: Point taken. I've removed the returns and changed them from bools to void.

File: Assets/Scripts/Game/PlayerActivate.cs (hunk 10 above)
Comment (BadLuckBurt) [reply Assets/Scripts/Game/PlayerActivate.cs]: I think it's useful to be able to directly use a Gameobject's name for that check but opinions may differ on that. If that check is ever changed or enhanced, it'll only need to be done once, decreasing the chance of errors.

File: Assets/Scripts/Utility/RDBLayout.cs (hunk 11 above)
Comment (BadLuckBurt) [reply Assets/Scripts/Utility/RDBLayout.cs]: Done

File: Assets/Scripts/Game/PlayerActivate.cs (hunk 9 above)
Comment (BadLuckBurt) [reply Assets/Scripts/Game/PlayerActivate.cs]: As stated above, point taken. Returns have been removed and method declaration changed from bool to void.
